Add plugin: Note Status

soler1212 opened the pull request:
# I am submitting a new Community Plugin

## Repo URL

<!--- Paste a link to your repo here for easy access -->
Link to my plugin: https://github.com/devonthesofa/obsidian-note-status

## Release Checklist
- [x] I have tested the plugin on
  - [x]  Windows
  - [ ]  macOS
  - [x]  Linux
  - [ ]  Android _(if applicable)_
  - [ ]  iOS _(if applicable)_
- [x] My GitHub release contains all required files (as individual files, not just in the source.zip / source.tar.gz)
  - [x] `main.js`
  - [x] `manifest.json`
  - [x] `styles.css` _(optional)_
- [x] GitHub release name matches the exact version number specified in my manifest.json (_**Note:** Use the exact version number, don't include a prefix `v`_)
- [x] The `id` in my `manifest.json` matches the `id` in the `community-plugins.json` file.
- [x] My README.md describes the plugin's purpose and provides clear usage instructions.
- [x] I have read the developer policies at https://docs.obsidian.md/Developer+policies, and have assessed my plugins's adherence to these policies.
- [x] I have read the tips in https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines and have self-reviewed my plugin to avoid these common pitfalls.
- [x] I have added a license in the LICENSE file.
- [x] My project respects and is compatible with the original license of any code from other plugins that I'm using.
      I have given proper attribution to these other projects in my `README.md`.

soler1212 commented:
I've decided to close the publication of this release. I want to submit the plugin to the plugins list with a more refined version. Soon, I’ll open a new PR with an updated release.

soler1212 commented:
New release 1.0.10 up and running with multiple improvements

joethei commented:
[Copyright (C) 2020-2025 Dynalist Inc.](https://github.com/devonthesofa/obsidian-note-status/blob/49630ccdcb7414030bdae93a66deda90a2f73da4/LICENSE#L3)
Update the license file.

["keywords": ["status", "notes", "organization", "productivity", "workflow"],](https://github.com/devonthesofa/obsidian-note-status/blob/49630ccdcb7414030bdae93a66deda90a2f73da4/manifest.json#L13-L19)
These manifest fields don't exist, remove them.

[name: 'Refresh Status',](https://github.com/devonthesofa/obsidian-note-status/blob/49630ccdcb7414030bdae93a66deda90a2f73da4/main.ts#L234), [name: 'Force Refresh UI',](https://github.com/devonthesofa/obsidian-note-status/blob/49630ccdcb7414030bdae93a66deda90a2f73da4/main.ts#L243), [name: 'Insert Status Metadata',](https://github.com/devonthesofa/obsidian-note-status/blob/49630ccdcb7414030bdae93a66deda90a2f73da4/main.ts#L251), [name: 'Open Status Pane',](https://github.com/devonthesofa/obsidian-note-status/blob/49630ccdcb7414030bdae93a66deda90a2f73da4/main.ts#L261), [name: 'Add Status to Current Note',](https://github.com/devonthesofa/obsidian-note-status/blob/49630ccdcb7414030bdae93a66deda90a2f73da4/main.ts#L268), [.setTitle('Change Status')](https://github.com/devonthesofa/obsidian-note-status/blob/49630ccdcb7414030bdae93a66deda90a2f73da4/main.ts#L296), [.setTitle('Change Status')](https://github.com/devonthesofa/obsidian-note-status/blob/49630ccdcb7414030bdae93a66deda90a2f73da4/main.ts#L323), [.setTitle('Change Note Status')](https://github.com/devonthesofa/obsidian-note-status/blob/49630ccdcb7414030bdae93a66deda90a2f73da4/main.ts#L339), [.setTitle('Change Note Status')](https://github.com/devonthesofa/obsidian-note-status/blob/49630ccdcb7414030bdae93a66deda90a2f73da4/main.ts#L339), [name: 'Colorful Workflow',](https://github.com/devonthesofa/obsidian-note-status/blob/49630ccdcb7414030bdae93a66deda90a2f73da4/constants/status-templates.ts#L19), [name: 'Minimal Workflow',](https://github.com/devonthesofa/obsidian-note-status/blob/49630ccdcb7414030bdae93a66deda90a2f73da4/constants/status-templates.ts#L35), [name: 'Academic Research',](https://github.com/devonthesofa/obsidian-note-status/blob/49630ccdcb7414030bdae93a66deda90a2f73da4/constants/status-templates.ts#L46), [name: 'Project Management',](https://github.com/devonthesofa/obsidian-note-status/blob/49630ccdcb7414030bdae93a66deda90a2f73da4/constants/status-templates.ts#L60), [text: 'Import Template',](https://github.com/devonthesofa/obsidian-note-status/blob/49630ccdcb7414030bdae93a66deda90a2f73da4/settings/settings-tab.ts#L307), [text: 'Export Templates',](https://github.com/devonthesofa/obsidian-note-status/blob/49630ccdcb7414030bdae93a66deda90a2f73da4/settings/settings-tab.ts#L318), [this.toolbarButton.setAttribute('aria-label', 'Note Status');](https://github.com/devonthesofa/obsidian-note-status/blob/49630ccdcb7414030bdae93a66deda90a2f73da4/ui/components/status-dropdown.ts#L83), [this.toolbarButton.setAttribute('aria-label', 'Note Status');](https://github.com/devonthesofa/obsidian-note-status/blob/49630ccdcb7414030bdae93a66deda90a2f73da4/ui/components/status-dropdown.ts#L498), [return 'Status Pane';](https://github.com/devonthesofa/obsidian-note-status/blob/49630ccdcb7414030bdae93a66deda90a2f73da4/ui/components/status-pane-view.ts#L27), [title: 'Refresh Statuses',](https://github.com/devonthesofa/obsidian-note-status/blob/49630ccdcb7414030bdae93a66deda90a2f73da4/ui/components/status-pane-view.ts#L129), [item.setTitle('Change Status')](https://github.com/devonthesofa/obsidian-note-status/blob/49630ccdcb7414030bdae93a66deda90a2f73da4/ui/components/status-pane-view.ts#L258), [item.setTitle('Open in New Tab')](https://github.com/devonthesofa/obsidian-note-status/blob/49630ccdcb7414030bdae93a66deda90a2f73da4/ui/components/status-pane-view.ts#L269)

[private updateExistingFrontmatter(](https://github.com/devonthesofa/obsidian-note-status/blob/49630ccdcb7414030bdae93a66deda90a2f73da4/services/status-service.ts#L199-L241), [export function updateFrontmatterProperty(](https://github.com/devonthesofa/obsidian-note-status/blob/49630ccdcb7414030bdae93a66deda90a2f73da4/utils/file-utils.ts#L42-L70)
Use the `processFrontMatter` function from the Obsidian API.

[css += `](https://github.com/devonthesofa/obsidian-note-status/blob/49630ccdcb7414030bdae93a66deda90a2f73da4/services/style-service.ts#L86-L157)
This should be part of the `styles.css` file.

[containerEl.createEl('h3', { text: 'UI Settings' });](https://github.com/devonthesofa/obsidian-note-status/blob/49630ccdcb7414030bdae93a66deda90a2f73da4/settings/settings-tab.ts#L25), [containerEl.createEl('h3', { text: 'Custom Statuses' });](https://github.com/devonthesofa/obsidian-note-status/blob/49630ccdcb7414030bdae93a66deda90a2f73da4/settings/settings-tab.ts#L120), [containerEl.createEl('h3', { text: 'Status Templates' });](https://github.com/devonthesofa/obsidian-note-status/blob/49630ccdcb7414030bdae93a66deda90a2f73da4/settings/settings-tab.ts#L233-L237)
For section headings in settings use:
```ts
new Setting(containerEl).setName('name here').setHeading();
```

[containerEl.createEl('h3', { text: 'UI Settings' });](https://github.com/devonthesofa/obsidian-note-status/blob/49630ccdcb7414030bdae93a66deda90a2f73da4/settings/settings-tab.ts#L25), [containerEl.createEl('h3', { text: 'Custom Statuses' });](https://github.com/devonthesofa/obsidian-note-status/blob/49630ccdcb7414030bdae93a66deda90a2f73da4/settings/settings-tab.ts#L120), [containerEl.createEl('h3', { text: 'Status Templates' });](https://github.com/devonthesofa/obsidian-note-status/blob/49630ccdcb7414030bdae93a66deda90a2f73da4/settings/settings-tab.ts#L233-L237)
[Avoid "settings" in settings headings](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Avoid+%22settings%22+in+settings+headings)

[// Add tooltip](https://github.com/devonthesofa/obsidian-note-status/blob/49630ccdcb7414030bdae93a66deda90a2f73da4/ui/components/status-bar.ts#L67-L69), [// Add tooltip to icon too](https://github.com/devonthesofa/obsidian-note-status/blob/49630ccdcb7414030bdae93a66deda90a2f73da4/ui/components/status-bar.ts#L77-L79), [// Add tooltip](https://github.com/devonthesofa/obsidian-note-status/blob/49630ccdcb7414030bdae93a66deda90a2f73da4/ui/components/status-bar.ts#L102-L104), [// Add tooltip to the chip element](https://github.com/devonthesofa/obsidian-note-status/blob/49630ccdcb7414030bdae93a66deda90a2f73da4/ui/components/status-dropdown-component.ts#L322-L324), [optionEl.setAttribute('aria-label', `${status.name} - ${status.description}`);](https://github.com/devonthesofa/obsidian-note-status/blob/49630ccdcb7414030bdae93a66deda90a2f73da4/ui/components/status-dropdown-component.ts#L459-L460)
use the `setTooltip` function for this.

[export function extractFrontmatter(content: string): {](https://github.com/devonthesofa/obsidian-note-status/blob/49630ccdcb7414030bdae93a66deda90a2f73da4/utils/file-utils.ts#L17-L36), [export async function getFrontmatterProperty(](https://github.com/devonthesofa/obsidian-note-status/blob/49630ccdcb7414030bdae93a66deda90a2f73da4/utils/file-utils.ts#L76-L95)
Use the MedataCache to read frontmatter.

soler1212 commented:
Thanks for the review @joethei . I've made the following fixes:

1. Updated the license file with correct copyright information.

2. Removed the "keywords" field from manifest.json as it's not a supported field.

3. Fixed naming consistency issues by using sentence case for UI elements as per Obsidian guidelines.

4. Replaced custom frontmatter manipulation with Obsidian's `processFrontMatter` API function.

5. Regarding dynamic CSS: I've moved static styles to styles.css, but kept dynamic color-related styles in JavaScript since they depend on user settings. If there's a better approach for handling dynamic styles, I'd appreciate guidance.

6. Updated settings headings to use `new Setting(containerEl).setName('name').setHeading()` instead of direct element creation.

7. Removed "settings" word from settings headings per guidelines.

8. Replaced manual tooltip implementation with the `setTooltip` function.

9. Regarding MetadataCache: I believe I'm already using it correctly, but I may have misunderstood the issue. I'm using `app.metadataCache.getFileCache(file)` to access frontmatter. Could you clarify if this approach is incorrect?

Let me know if anything needs further adjustment.

soler1212 commented:
/skip

The flagged style assignments in JavaScript are necessary because they handle dynamic values that cannot be predetermined in static CSS:

1. Status colors are user-customizable in settings and must be applied dynamically based on user preferences
2. The dropdown positioning system adapts to viewport boundaries and user interaction points (click positions)
3. Dynamic theme integration requires runtime calculation of CSS variables from Obsidian's theming system

These specific use cases require JavaScript-based style manipulation. All static styling has already been properly moved to the styles.css file.

The plugin follows established Obsidian plugin patterns for dynamic styling with CSS variables, which is why elements use `style.setProperty()` with CSS variables rather than direct style properties, maintaining the ability for themes to override values.
